registration/manager_test.go

```diff
@@ -141,7 +143,140 @@ func TestSelfRegistrationManager_defaultHTTPCheck(t *testing.T) {
+		mockClient.On("RegisterService", mock.Anything,
+			mock.MatchedBy(func(r *consulapi.AgentServiceRegistration) bool {
+				return r.ID == id && r.Name == defaultServiceName
+			})).Return(nil)
+		mockClient.On("DeregisterService", mock.Anything,
+			mock.MatchedBy(func(actual string) bool {
+				return actual == id
+			})).Return(nil)
+		manager.client = mockClient
+
+		ctx, cancel := context.WithCancel(context.Background())
+		errCh := make(chan error)
+		go func() {
+			if err := manager.Start(ctx); err != nil {
+				errCh <- err
+			}
+		}()
+		cancel()
+		select {
+		case err := <-errCh:
+			// Confirm that exit is due to context cancel
+			assert.Equal(t, err, context.Canceled)
+		case <-time.After(time.Second * 5):
+			t.Fatal("Run did not exit properly from cancelling context")
+		}
+		mockClient.AssertExpectations(t)
+	})
+
+	t.Run("error on register", func(t *testing.T) {
+		mockClient := new(mocks.ConsulClientInterface)
+		mockClient.On("RegisterService", mock.Anything, mock.Anything).
+			Return(errors.New("mock register error"))
+		manager.client = mockClient
+
+		ctx, cancel := context.WithCancel(context.Background())
+		defer cancel()
+		errCh := make(chan error)
+		go func() {
+			if err := manager.Start(ctx); err != nil {
+				errCh <- err
+			}
+		}()
+		select {
+		case err := <-errCh:
+			// Confirm that exit is due to mock error
+			assert.Contains(t, err.Error(), "mock register error")
+		case <-time.After(time.Second * 5):
+			t.Fatal("Run did not exit properly from cancelling context")
+		}
+		mockClient.AssertExpectations(t)
+	})
+
+	t.Run("error on deregister", func(t *testing.T) {
+		mockClient := new(mocks.ConsulClientInterface)
+		mockClient.On("RegisterService", mock.Anything, mock.Anything).
+			Return(nil)
+		mockClient.On("DeregisterService", mock.Anything, mock.Anything).
+			Return(errors.New("mock deregister error"))
+		manager.client = mockClient
+
+		ctx, cancel := context.WithCancel(context.Background())
+		errCh := make(chan error)
+		go func() {
+			if err := manager.Start(ctx); err != nil {
+				errCh <- err
+			}
+		}()
+		cancel() // cancel to initiate deregister
+
+		select {
+		case err := <-errCh:
+			// Confirm that exit is due to mock error
+			assert.Contains(t, err.Error(), "mock deregister error")
+		case <-time.After(time.Second * 5):
+			t.Fatal("Run did not exit properly from cancelling context")
+		}
+		mockClient.AssertExpectations(t)
+	})
+}
+
+func TestSelfRegistrationManager_Deregister(t *testing.T) {
+	t.Parallel()
+	id := "cts-123"
+	ctx := context.Background()
+	manager := &SelfRegistrationManager{
+		service: &service{
+			name: defaultServiceName,
+			id:   id,
+		},
+		logger: logging.NewNullLogger(),
+		// mock client will be set per test case
+	}
+
+	t.Run("success", func(t *testing.T) {
+		mockClient := new(mocks.ConsulClientInterface)
+		mockClient.On("DeregisterService", mock.Anything,
+			mock.MatchedBy(func(actual string) bool {
+				return actual == id
+			})).Return(nil)
```

Comment:
```suggestion
		mockClient.On("DeregisterService", mock.Anything, id).Return(nil)
```

Reply:
Oh, sorry I misread your suggestion! That makes sense to me, `MatchedBy` would be better for when you don't need an exact match but an approximate custom match instead.

Reply:
Actually, even passing ctx as the first arg would work.
i.e.
`mockClient.On("DeregisterService", ctx, id).Return(nil)`

Reply:
Will update other tests with this format where relevant!